core/src/main/java/hudson/model/AbstractCIBase.java

```diff
@@ -158,21 +160,31 @@ private void updateComputer(Node n, Map<String, Computer> byNameMap, Set<Compute
     @CheckForNull
     private Computer createNewComputerForNode(Node n, boolean automaticAgentLaunch) {
         Computer c = null;
-        Map<Node, Computer> computers = getComputerMap();
+        ConcurrentMap<Node, Computer> computers = getComputerMap();
         // we always need Computer for the built-in node as a fallback in case there's no other Computer.
         if (n.getNumExecutors() > 0 || n == Jenkins.get()) {
+            // The start/connect of a new computer is potentially costly, so we don't want to perform it inside
+            // computeIfAbsent. Instead, we use this creationWasAttempted flag to signal if start/connect is needed or
+            // not.
+            final AtomicBoolean creationWasAttempted = new AtomicBoolean(false);
             try {
-                c = n.createComputer();
+                c = computers.computeIfAbsent(n, node -> {
+                    creationWasAttempted.set(true);
+                    return node.createComputer();
+                });
             } catch (RuntimeException ex) { // Just in case there is a bogus extension
                 LOGGER.log(Level.WARNING, "Error retrieving computer for node " + n.getNodeName() + ", continuing", ex);
             }
+            if (!creationWasAttempted.get()) {
+                LOGGER.log(Level.WARNING, "Node {0} is not a new node skipping", n.getNodeName());
```

Comment:
Optional non-blocking feedback for your consideration: In the scenario described in [JENKINS-69534](https://issues.jenkins.io/browse/JENKINS-69534), where in thread 1 a `ComputerA1` for `NodeA` fails to launch and `KubernetesLauncher#launch(SlaveComputer, TaskListener)` invokes `AbstractCloudSlave#terminate`, and in thread 2 a timer thread invokes `NodeProvisioner#suggestReviewNow`, resulting in both threads entering `AbstractCIBase#createNewComputerForNode(Node, boolean)` for `NodeB`, the losing thread will log this statement at warning level. Since this behavior is expected, it seems to me that warning level might be too high a level for this log statement, being better reserved for unexpected but non-fatal scenarios. Perhaps a debug or fine log would be more appropriate if this is the expected behavior. On the other hand, if you think this is exceptional enough to warrant a warning, then feel free to leave this code as-is.

Reply:
Thanks!
This log line was taken directly from line `219` of the previous version of `AbstractCIBase` and I didn't reflect on the level when I moved it. I agree that `WARNING` is too high, I'll change it to `FINE`.